### PR TITLE
Add features to user interface

### PR DESCRIPTION
**Pull request by hughesbm**
- Games can now be created via front-end
- Games are now automatically populated when created via front-end
- Pieces are now correctly assigned to the black player when they join the game (via front-end)
- Users cannot join games they are already playing in
- Index now shows games awaiting players at the top of the page

### CODE REVIEW

File: app/controllers/games_controller.rb [hunk 1]
```diff
@@ -1,32 +1,30 @@
 class GamesController < ApplicationController
-  before_action :set_game, only: [:show, :update]
+  before_action :authenticate_user!, only: [:create, :update, :forfeit]
+  before_action :find_current_game, only: [:show, :update, :forfeit]
+  before_action :prevent_user_from_joining_own_game, only: :update
 
   def index
-    @games = Game.all.order('updated_at').reverse_order
-    # @games = Game.where('white_player_id is NULL OR black_player_id is NULL').order('created_at').reverse_order
+    @games = Game::GameCategorizer.new(Game.all)
```
Comment (mscoutermarsh) [inline app/controllers/games_controller.rb]: :+1:

I'd name this something like... `Games::IndexPresenter`. Which i think is a bit more clear about what it does

File: app/services/game/game_categorizer.rb [hunk 2]
```diff
@@ -0,0 +1,17 @@
+class Game::GameCategorizer
+  def initialize(games)
+    @all = games
```
Comment (mscoutermarsh) [inline app/services/game/game_categorizer.rb]: Style thing that I personally do.

`@games = games`

Then above the initializer `attr_reader :games`

Then within this object, reference `games` instead of `@games`.

Only reasoning I do this is. `@games` is never updated outside the initializer. So I think having only a reader makes it really clear

Comment (mscoutermarsh) [inline app/services/game/game_categorizer.rb]: instance vars make me nervous. State that can change. Makes things hard to follow. (again, really just personal opinion, up to what you think)

File: app/controllers/games_controller.rb [hunk 3]
```diff
@@ -1,32 +1,30 @@
...
   end
 
-  def new
-    @game = Game.new
-  end
-
   def create
-    @game = Game.create(game_params)
+    @game = Game.create(game_params.merge(white_player_id: current_user.id))
+    @game.populate_game!
```
Comment (mscoutermarsh) [inline app/controllers/games_controller.rb]: Could remove `populate_game` from the model entirely and call your object directly from here.

Then for tests, could have them only be for `/app/services/games/board_populator_spec.rb` (something like that). Which gets the tests out of `Game`

File: app/services/game/game_categorizer.rb (hunk 2 above)
Comment (hughesbm) [inline app/services/game/game_categorizer.rb]: Thanks, that's really helpful. I've read about avoiding instance variables in these situations but I didn't connect the dots about why it's important or how to do it.
